### DISCUSSION TITLE
Slow GeoShapeQuery Performance in Bleve (v2.4.3)

### DISCUSSION THREAD

**Discussion by MeonAigeanta** [blevesearch/bleve#2114] [Q&A] [answered]
I'm experiencing slow performance with Bleve's GeoShapeQuery in my Go (v1.22) application. The query takes around 355 ms to return 51,092 hits, and I'm hoping to optimize it for faster results and understand why my code is running slow.

I'm using the following code to create the GeoShapeQuery:
```
geojson := `{"type":"Polygon","coordinates":[[[-152,62],[-152,59],[-147,59],[-147,62],[-152,62]]]}`
shape, _ := geo.ParseGeoJSONShape([]byte(geojson))
gsq := &query.GeoShapeQuery{
    FieldVal: "geog",
    Geometry: query.Geometry{
      Shape:    shape,
      Relation: "within",
    },
  }
sreq := bleve.NewSearchRequest(gsq)
```
I varied the batch size used to build the index then searched with one of the polygons:

Batch | > 1000 | > 500 | > 100 | > 50 | No batching
-- | -- | -- | -- | -- | --
ms | 683 | 492 | 355 | 351 | 364

I non-randomly varied the number of entries (batch > 100): 

Entries | 100,000 | 200,000 | 300,000 | 400,000 | 526,286
-- | -- | -- | -- | -- | --
ms | 91 | 139 | 197 | 261 | 373

I varied the size of the polygon (batch > 100):

polygon | 1 | 2 | 3 | 4 | 5
-- | -- | -- | -- | -- | --
ms | 354 | 366 | 390 | 372 | 327
hits | 51,092 | 35,812 | 12,474 | 363 | 6

Setting the relation to “within” is faster than using “intersects”.  

I'd appreciate any guidance on optimizing my GeoShapeQuery performance. Are there any potential issues with my code or configuration that could be causing the slow results?

I've created a complete minimal example on github: 
https://github.com/MeonAigeanta/bleve-search-test/commit/a7d782c84e994c436799b8a390e070e3c3158533

**Comment by abhinavdangeti** [comment]
Hey @Likith101 would you look into this one when you can please.

**Comment by Likith101** [comment, marked as answer]
Hey @MeonAigeanta , apologies for the late response.

I went through your code and I only find one thing that could help improve your performance and that is multithreading the queries. For a use case with one query at a time however, this is the maximum as of now.

There are a few work in progress commits to improve performance in this area as of now. You could perhaps try cherry picking this (https://github.com/blevesearch/zapx/pull/258) commit on a local bleve/zapx build and run the performance numbers again.

**Reply by MeonAigeanta** [reply]
Thank you for looking at my code and for your suggestions.
